### ISSUE TITLE
Proposal to reserve "version" member at top level of resource objects

### ISSUE THREAD

**Issue by scott-mcdonald** [json-api/json-api#600]
Currently each resource has "identity" via the "type" and "id" top-level keywords at the top-level of a resource, but a resource can be at a different "version" based on updates, etc.

I propose adding an _optional_ top-level resource keyword called **version** that is an opaque identifier/value (read that as string) of the current version of the respective individual resource. You can then use "type", "id", and "version" to uniquely identify a resource and at what version this representation is. I believe having the version be a known property on a json-api compliant resource further refines the "identity" of the resource and makes certain client-side workflows easier, i.e. optimistic locking, client-side caching, etc.

**Comment by hhware** [comment]
@scott-mcdonald, I contemplated proposing this as well a while ago, but I dropped that idea at that time because I thought that it would not fit situations when another existing attribute was being used for version, so that naming it `version` would not be always appropriate. However, with the advent of JSON-API-specific top level of resource objects (#588), it becomes an attractive option again: it would not look awkward if `version` were populated with a value of an existing attribute. E.g., in the example below `id` is mapped to `username`, and `version` is mapped to `updated_at`:

```
{
  "data": {
    "id": "jsmith",
    "type": "users",
    "version": "2014-10-10T14:13:24.000Z",
    "attributes": {
      "username": "jsmith",
      "first_name": "john",
      "last_name": "smith",
      "updated_at": "2014-10-10T14:13:24.000Z"
    }
  }
}
```

I think that standardizing on a top level resource member for the purposes of versioning of resources would be perfectly in line with the anti-bikeshedding impetus of the spec. Having such member does not seem to anyhow conflict with supporting `ETag` at the same time and will be in fact necessary when compound documents are returned. I support this proposal.

**Comment by tkellen** [comment]
Having a standardized place to put a version isn't enough to generalize this approach. We'd also have to take a stance on what the version string could contain if we hoped for a generic client to be able to reason about what is considered "newer". At the moment, this feels like an over-reach. That said, I'm +1 to considering this more carefully after 1.0.

**Comment by cziegenberg** [comment]
I'd also like to have something like a version. I thought about implementing it as an attribute or meta field, but standarizing it makes sense... +1

BTW: A UTC date time value is IMHO a very good base for a version, and in most cases it already exists for the resource. And it should also include milliseconds (which can default to 000 if not available) like in the example. I had to fix a nasty bug in a web service today, caused by multiple updates sent in the same second - so I can confirm that this can be really important. :)

**Comment by scott-mcdonald** [comment]
I would argue the opposite - you would **not** want or need _json-api_ to standardize the content of the version string (treat it as an opaque value up to the service to define in their own specification) because _fundamentally_ all a client and/or service wants to know is if two resources with the same type/id are the same version or not.

For any client, this is easy by using string equivalence between 2 string values - are the string values equivalent or not - i.e. are the versions the same or not.

For an individual service, they define the content/specification of the version. For example the service developer could design it to be a GUID, a date/time, a ROWVERSION column in a database, or whatever - up to the service provider as this is implementation details. The service then serializes the actual type to a string and the client treats version string as an immutable, opaque value as defined previously. On updates, the client should pass the received opaque version string to the service on any update operation if a version was provided on the read operation.

**Comment by cziegenberg** [comment]
> fundamentally all a client and/or service wants to know is if two resources with the same type/id are the same version or not

That's right, to detect and deny update requests that rely on old data, this would be enough.

**Comment by dgeb** [comment]
Specific versioning schemes could be implemented in extensions. There is simply no way we're going to agree upon a single versioning scheme for resources that works well across a wide variety of distributed systems.

**Comment by hhware** [comment]
> We'd also have to take a stance on what the version string could contain if we hoped for a generic client to be able to reason about what is considered "newer".

@tkellen, JSON API does not standardize URL format, which seems to have much wider impact. Following that approach, I think that standardizing location of a version string is as far as JSON API could possibly go.

**Comment by tkellen** [comment]
I'm not sure what you mean about our stance on URLs. That said, if the usage of the version value is implementation specific, it may as well just live in `meta`, no?

**Comment by scott-mcdonald** [comment]
Version could be in the _meta_, _attributes_, or _extensions_, true. Personally, I feel it is a part of resource identity _at any moment in time_ as resources mutate over time - in my mind I think of resource identity as the following in order:
1. type
2. id
3. version 
and hence my proposal of version as a first-class property.

With that said, I like and have chosen _json-api_ over some of the other JSON standards like HAL and Siren. I am currently adopting it as the standard for my own REST API - excellent work gentlemen.

**Comment by hhware** [comment]
@tkellen, perhaps my example was indeed not the best, still, I agree with @scott-mcdonald that version makes sense as part of resource identifier even in the absence of standard algorithm of comparing versions. Several versions of the same resource can be active and be used by API client at the same time (e.g., one "published" and one currently being updated to be published soon), and `version` member can be used, for instance, to build cache key for versioned resources.

**Comment by tkellen** [comment]
I agree that the base specification should provide guidance here. I am milestoning this for consideration in 1.1.

**Comment by tkellen** [comment]
@dgeb and I just discussed this and we are :+1: to add this to 1.1 alpha pending thoughts from @ethanresnick and @steveklabnik.

**Comment by steveklabnik** [comment]
> the current version of the respective individual resource

I have tons of thoughts, but not the time to type them out right now. I think this adds _waaaay_ more complexity for little real gain, you should really version your entire API, not individual endpoints.

**Comment by tkellen** [comment]
We're not talking about versioning endpoints, we're talking about versioning resource objects. Think etags.

**Comment by steveklabnik** [comment]
Sorry, I think the "real REST" vs "not real REST" terminology is messing me up here.

I am incredibly skeptical, but I may just be an old man here...

**Comment by tkellen** [comment]
No worries! I don't think you're being an old man... you being skeptical makes me skeptical too. Let's chat about this on IRC and see if we can figure out what the disconnect is.

**Comment by scott-mcdonald** [comment]
FYI, it is not a "schema" version but an "instance" or "object" version of the resource itself. I currently use version in Attributes of resources and use it for optimistic locking purposes in terms of "updating" the resource and caching purposes.

**Comment by cziegenberg** [comment]
I'm currently working on an ORM implementing a versioning behavior and would love to use versioning in the JSON API representation of the resources.

Optimistic locking ist the main reason for the version, but I also implemented a history for the changes between versions, and I plan to implement the ability to switch to a given previous version. Another interesting function is the automatic merging of changes, when data for an old version have been sent.

Just to give you an idea about what's also possible with such a version... and why a normal attribute perhaps isn't enough (special errors and responses, special actions required e.g. to confirm suggested merge data,...).

**Comment by tkellen** [comment]
`<bikeshed>`
I think we should call it a `fingerprint`, not `version`. We already use the `version` member in the specification and I don't want to have to differentiate between them.
`</bikeshed>`

**Comment by steveklabnik** [comment]
After chatting with @tkellen , I think the 'version' name is a red herring that scared me off. I'm generally in favor of adding something like this, I just think the name is too overloaded.

**Comment by dgeb** [comment]
Clearly overloading of `version` can get confusing. `fingerprint` is probably better, but to my ear it sounds like a checksum (maybe that's only me though).

@ziege do you have any requirements that would not be met by a string value?

**Comment by tkellen** [comment]
I think of a checksum as being a specific form of fingerprint, used for tamper-proofing. I don't much care what we call it though, as long as it isn't `version`.

**Comment by dgeb** [comment]
Maybe `revision`?

**Comment by tkellen** [comment]
perfect.

**Comment by scott-mcdonald** [comment]
`revision` +1

**Comment by ethanresnick** [comment]
For single-resource requests, my feeling is that this would just be duplicating Etags in a non-standard way, with little benefit over using Etags directly. Right now, I can do a `GET /people/1`, and the Etag I get back is conceptually identical to what the `revision` key would hold. Then, if I want to `PATCH /people/1` and prevent unintended update overwriting, I provide an `If-Match` header with the value of the Etag I got before. (And, if the client doesn't put an `If-Match`, the server can respond `428 Precondition Required`.)

It's true that `If-Match` doesn't give the server the option to accept updates that the client's made to an out-of-date version and then merge those changes into the current version. So I guess that's one advantage a `revision` key would have. Also, I think we may increasingly see people using the JSON API JSON format over non-HTTP protocols (esp. web sockets), so having some way to embed this info in the JSON could be useful for that.

The most interesting scenario, I think, is when this is used in requests for a collection, like `GET /people`. Then each resource can get it's own `revision`, whereas an Etag can only apply to the whole response. That could be useful. 

However, if that's the primary use case, I think we should consider requiring that the `revision` key, or whatever it's called, MUST match the Etag that would be returned on a request to that resource object's `self` URI. That way, clients could send conditional-GET requests to update individual resources that they previously downloaded as part of a larger collection; if `revision` isn't coordinated with Etag, then this becomes impossible unless we invent some new mechanism just for that.

**Comment by tkellen** [comment]
@ethanresnick Your response basically traces the outline of the conversation I had with @steveklabnik yesterday. I'm :+1: to normatively requiring `revision` values to match etag values when HTTP is being used for transport.

**Comment by ethanresnick** [comment]
@tkellen Ok, cool!

Next question, then: should we allow `revision` as a top-level property on a collection request too? E.g.

```
{
  "data": [some resources],
  "revision": "[opaque string]"
}
```

I'm not sure what a concrete use case for that would be, but it seems consistent/like a logical extension of the above. Does it cause any problems?

**Comment by dgeb** [comment]
I have a problem requiring ETag / revision equivalency. I believe ETags should apply to response documents, while revisions should apply to each resource represented in that document.

I would expect different ETags when requesting `people/1` and `people/1?fields[people]=name`, even if person 1 doesn't change at all between requests. In the same scenario, I would expect `revision` for the person resource to remain the same.

On the other hand, `revision` could change even if all other aspects of a resource remain constant (e.g. if an attribute is changed to A, then B, then back to A again).

**Comment by ethanresnick** [comment]
> I would expect different ETags when requesting people/1 and people/1?fields[people]=name, even if person 1 doesn't change at all between requests

@dgeb Why? Is there anything in HTTP that implies that those two Etags need to be different? 

As I'm imagining it, an API that's tracking `revision` could just drop that value into the Etag

> On the other hand, revision could change even if all other aspects of a resource remain constant (e.g. if an attribute is changed to A, then B, then back to A again).

Yeah, that's true. But would it be a big problem if the Etag changed here too? Given that the `revision` number, which is included in the JSON body, would change, I think the Etag would need to change too.

**Comment by scott-mcdonald** [comment]
If I may, I would like to add the following comments:

**First**, I believe Ethan is spot on that `revision` and `etag` are one and the same. This was my original point when I first proposed this idea in this thread about 2 months ago of `version` being an _opaque identifier_ that the web server assigns for a specific version of a resource so that either a client or server can simply compare for equality to know whether two representations of a resource are the same - my main purpose for this was for _optimistic concurrency control_ on resource updates. This definition is verbatim what the `ETag` semantic meaning is in the HTTP protocol. My main goal was to make this a first-class property of a resource instead of being in Attributes or Meta or the HTTP header, part of the json:api standard, etc. I liked it being an _opaque identifier_ to maximize flexibility to let the server decide what exactly it is and not the json:api standard. Some people in this thread wanted a time-stamp like value, for my company we basically use ROWVERSION in a SQL Server database as the version of the resource - our particular implementation detail. Again per the HTTP specification for `ETag` let the web server "decide".

Therefore with accepting the fact that `revision` and `etag` are the same, his other comment's both for and against were spot on with the biggest **FOR** being for a collection of resources each resource getting its own individual `revision` value which you can not do with _etag_ at the HTTP header level - a win. Also applies for any "included" resources for a compound document.

**Second**, I would like to propose an alternative/enhancement which makes my original proposal "more RESTful" and just see what others think. This is a rhetorical statement but REST fully embraces HTTP as it's protocol so I think json:api should do the same in the following way:

Instead of deciding on a different name for essentially `etag` like `version` or `revision` or `fingerprint`, etc. fully embrace the current HTTP standard for all of these concepts/definitions for _optimistic concurrency control_ and _caching_ by putting the equivalent HTTP names in the `revision` node so `revision` now becomes a top level bag/collection/container like `attributes` for an individual resource. The json:api standard would then be using the exact same names with the exact same semantic meanings in the HTTP standard that everybody already knows and understands, etc. Makes adoption, understanding, and documentation easier, etc.

For example something like this:

```
{
  "links": {
    "self": "http://example.com/posts",
    "next": "http://example.com/posts?page[offset]=2",
    "last": "http://example.com/posts?page[offset]=10"
  },
  "data": [{
    "type": "posts",
    "id": "1",
    "revision": {
        last-modified: "Fri, 31 Jul 2015 20:29:32 GMT",
        max-age: 120,
        etag: "123456789"
    },
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    },
    "links": {
      "self": "http://example.com/posts/1"
    }
  }, {
    "type": "posts",
    "id": "2",
    "revision": {
        last-modified: "Sat, 01 Aug 2015 09:10:42 GMT",
        max-age: 360,
        etag: "987654321"
    },
    "attributes": {
      "title": "JSON API paints my garage!"
    },
    "links": {
      "self": "http://example.com/posts/2"
    }
  }]
}
```

And to play off of Ethan's idea about the revision node being at the "collection" level, it would contain default values that would apply for any resources in that collection unless overridden by an individual resource, for example a default `max-age` for all resources in that collection, etc.

Or something along this train of thought...

**Comment by hhware** [comment]
> fully embrace the current HTTP standard for all of these concepts/definitions for optimistic concurrency control and caching by putting the equivalent HTTP names in the revision node so revision now becomes a top level bag/collection/container like attributes for an individual resource.

IMHO, a great proposal!

**Comment by ethanresnick** [comment]
>  Also applies for any "included" resources for a compound document.

Crap. I forgot about this. Imagine a request for `/people/1?include=children`. If the ETag on that response matches the `revision` of the primary data (the `(people, 1)` resource), then, if one of the included `children` resources is updated, the Etag needs to be updated for HTTP caching, even though the `revision` on the primary resource hasn't changed. So there goes tying this to ETag—at least in all cases. Dan called it :)

I need to think about this more now...

**EDIT**: I'm thinking we could still put _some_ requirements on ETag–`revision` correspondence, to enable the conditional GET I described above. Because if we don't give that to clients, it seems like we're missing out on one big possibility for this feature. Then again, a response can have default includes, such that even `GET /people/1` might need to change `ETag`s even while the `(people, 1)` revision remains constant. I guess a correspondence based on `GET /people/1?include=` would work, but that's pretty nasty.

**Comment by ethanresnick** [comment]
@scott-mcdonald Re your `revision` bag proposal... I'm not sure I'm comfortable tying things that closely to HTTP, in terms of the naming (e.g. `etag`) and the types of values that would end up there (e.g. `max-age`, which doesn't make sense on "push" protocols).

**Comment by scott-mcdonald** [comment]
I see the point @ethanresnick and @dgeb are making on `revision` and `etag` and their possible impedance mismatch - they are not exactly the same but may be the same depending on the scenario.

`etag` is at the header level and thus applies to the document as a whole. When the document is a single resource with no other resources they align nicely, with json:api documents this is not always the case.

`revision` is at the resource level by definition.

A json:api compliant document could be a single resource or a collection of resources with the possibility of it also being a heterogeneous compound document of resources when resources are in the "included" section.

I am going to go back to my original point of this proposal in that the identity of an individual resource is:
- `type` (required string)
- `id` (required string)
- `revision` (optional opaque string)

where `revision` facilitates optimistic locking and caching at the fine-grained resource level for clients that want to leverage this when using json:api. This works nicely when the json:api document is a single resource or collection of resources or a heterogeneous document of resources when resources are "included" and is actually needed if resource level optimistic concurrency control or caching is desired.

I would say the only requirement for `etag` and `revision` equivalence is when the json:api document is a single resource with no included resources - still a fairly common scenario to provide benefit to make this a requirement?

**Comment by dgeb** [comment]
> I would say the only requirement for etag and revision equivalence is when the json:api document is a single resource with no included resources - still a fairly common scenario to provide benefit to make this a requirement?

Considering that any document can include top-level `meta` data (or default included resources as mentioned above) that could vary and legitimately factor into its ETag value, I consider any kind of hard requirement for ETag / revision equivalence to be a stretch.

**Comment by scott-mcdonald** [comment]
> I consider any kind of hard requirement for ETag / revision equivalence to be a stretch.

_Agreed_. Should be up to the web server (read implementation detail) on how to handle etag/revision coalescing.

**Comment by tkellen** [comment]
Great points everyone!  @scott-mcdonald Would you please submit a PR taking a stab at actual normative text for this addition?

**Comment by scott-mcdonald** [comment]
Sure but can only do this in my off-time from work, mix in wife and kids so please give me a couple of days for an initial draft...

To summarize this proposal so we are all on the same page:
- Part of "Resource Identifier Object"
  - `type` (string required)
  - `id` (string required)
  - **`revision` (string optional)**
- Opaque identifier up to web server to define value.
- If supplied in a GET call, **MUST** be passed in a PATCH or PUT call.
- Any `etag` to `revision` coalescing is an implementation detail of the web server. Desirable if possible to increase HTTP caching mechanisms but not required.

Missing anything?

**Comment by tkellen** [comment]
No worries @scott-mcdonald, we're all juggling this in our free time for the most part. That looks good except that we cannot say that `PATCH` or `PUT` **MUST** include `revision`--it would be backwards incompatible with `1.0` of the spec. This change must be wholly additive.

**Comment by dnperfors** [comment]
As I read through this topic and the draft for this revision field, I only see that it will be used for resources. Isn't it needed/useful for relationships as well?

**Comment by wimleers** [comment]
FWIW: Drupal has been calling this "revisions" for more than a decade.

We're currently working on exposing revisions of resources ("entities" in Drupal terminology) via JSON API: https://www.drupal.org/project/jsonapi/issues/2992833. Of course we'd much prefer it if the JSON API spec explicitly documented this, rather than coming up with our own implementation.

**Comment by scott-mcdonald** [comment]
@wimleers Agreed. This enhancement was targeted to be included in the 1.1 version of the specification, I did write up a pull request #824 that surprisingly does not have any conflicts considering I wrote it back in July of 2015. At our company, we have added `revision` as an attribute but would have preferred it to be in the spec. I believe it will get there under the guise of "eventual consistency" ;-)

**Comment by ethanresnick** [comment]
@wimleers @scott-mcdonald I'd love to standardize this at some point but, for now, since profiles are finally done, I think they could offer the perfect way to experiment with resource versioning

**Comment by wimleers** [comment]
+1, but profiles did not _ship_ yet. Which brings me to an important question: why not ship a `1.1` release _right now_, so that such experiments become possible? :)

**Comment by ethanresnick** [comment]
@wimleers It's gonna happen very soon. I'm just working on a PR for profile registration, and I think #1304, #1302 and #1199 should get in too. The way I think of beta, though, is that it's like [TC39's Stage 3](https://tc39.github.io/process-document/). I.e., people should be trying to implement the new features once they get to beta stage (even if not shipping them yet), just to make sure that implementation isn't unexpectedly difficult before we finalize the designs for good.

**Comment by wimleers** [comment]
> I.e., people should be trying to implement the new features once they get to beta stage (even if not shipping them yet), just to make sure that implementation isn't unexpectedly difficult before we finalize the designs for good.

This is good to know. I didn't realize that this was the case. Perhaps it's worth sending out a tweet from the `@jsonapi` Twitter account to announce that spec implementors should attempt to implement 1.1 in its current state and provide feedback?

**Comment by wimleers** [comment]
Now also cross-posted https://github.com/json-api/json-api/issues/600#issuecomment-429380085 to https://www.drupal.org/project/jsonapi/issues/2955020#comment-12818094.

**Comment by ethanresnick** [comment]
> Perhaps it's worth sending out a tweet from the @jsonapi Twitter account to announce that spec implementors should attempt to implement 1.1 in its current state and provide feedback?

100% agree. @dgeb can you do this if you think it's a good idea? I think you control that twitter account, right?

**Comment by dgeb** [comment]
@ethanresnick yep, I am pulling the strings behind the scenes on Twitter (not exactly expertly, I might add).

We did tweet about profiles landing for 1.1, but we have not explicitly asked that they be tried out. I agree it's probably a good idea as each major feature lands to do so 👍 

Thanks for the suggestion @wimleers!

https://twitter.com/jsonapi/status/1052647059982077953

**Comment by wimleers** [comment]
Retweeted! 💯
